src/models/helpers/isFlagged.ts

```diff
@@ -0,0 +1,626 @@
+ * @param {object} model - The Sequelize model for which to retrieve the foiaable columns.
+ * @returns {Promise<Array<object>>} - A promise that resolves to an array of objects representing
+ * the foiaable columns.
+ */
+const getFoiaableColumnsForModel = async (
+  models,
+  model,
+) => models.Foiaable.findAll({
+  attributes: ['column'], // Selects only the 'column' attribute from the result set
+  // Filters the result set based on the 'table' attribute matching the model's table name
+  where: { table: model.getTableName() },
+  raw: true, // Returns raw data instead of Sequelize instances
+});
+
+/**
+ * Checks if there is an attempt to change foiaable columns on a model instance.
+ * @param sequelize - The Sequelize instance.
+ * @param instance - The model instance being checked.
+ * @param options - Additional options for the check.
+ */
+const checkForAttemptToChangeFoiaableValue = async (
+  sequelize,
+  instance,
+  options,
+) => {
+  // Check if the instance has the 'isFoiaable' attribute set to true
+  if (instance.get('isFoiaable')) {
+    // Get the list of foiaable columns for the model
+    const foiaableColumns = await getFoiaableColumnsForModel(
+      sequelize.models,
+      instance.constructor,
+    );
+
+    // Get the list of updated columns on the instance
+    const updatedColumns = instance.changed();
+
+    // Find any updated column that matches a foiaable column
+    const updatedFoiaableColumns = updatedColumns
+      .find((column) => foiaableColumns.includes(column));
+
+    // If there are updated foiaable columns, throw an error
+    if (updatedFoiaableColumns && updatedFoiaableColumns.length > 0) {
+      throw new Error(`Cannot update foiaable columns (${updatedFoiaableColumns.join(',')}) on table ${instance.constructor.getTableName()}`);
+    }
+  }
+};
+
+/**
+ * Checks if a record is foiaable and throws an error if it is.
+ * @param sequelize - The Sequelize instance.
+ * @param instance - The instance of the model.
+ * @param options - The options for the operation.
+ */
+const checkForAttemptToRemoveFoiaableValue = async (
+  sequelize,
+  instance,
+  options,
+) => {
+  // Check if the record is foiaable
+  if (instance.get('isFoiaable')) {
+    // Throw an error if the record is foiaable
+    throw new Error(`Cannot remove foiaable record in table ${instance.constructor.getTableName()}`);
+  }
+};
+
+//----------------------------------------------------
+
+/**
+ * This function is responsible for auto-populating a flag in an instance of a Sequelize model.
+ * It checks if the 'autoPopulateFlag' option is set to true, and if not, it sets the flag to
+ * false and adds it to the list of fields to be saved.
+ *
+ * @param sequelize - The Sequelize instance.
+ * @param instance - The instance of the Sequelize model.
+ * @param options - The options object.
+ * @param flagName - The name of the flag to be auto-populated.
+ */
+const autoPopulateFlag = (
```

Comment:
this is nearly identical to the function at the top of this file, autoPopulateIsFlagged, if we need the check at the top of this function, maybe condense so we aren't repeating so much

Reply:
your right, I have been working on this for too long. Thanks I will reduce and reuse